controllers/lighting_controller/lighting_controller.py

```diff
@@ -0,0 +1,270 @@
+"""
+The controller for altering arena lighting provided by a DirectionalLight and a Background
+Currently doesn't support:
+- Timed pre-match lighting changes
+"""
+import sys
+from typing import Dict, List, Tuple, Optional, NamedTuple
+from pathlib import Path
+from dataclasses import dataclass
+
+from controller import Node, Supervisor
+
+REPO_ROOT = Path(__file__).resolve().parent.parent.parent
+
+sys.path.insert(1, str(REPO_ROOT / 'modules'))
+
+import controller_utils  # isort:skip
+import webots_utils  # isort:skip
+
+
+class ArenaLighting(NamedTuple):
+    light_def: str
+    intensity: float = 3.5
+    colour: Tuple[float, float, float] = (1, 1, 1)
+
+
+class LightingEffect(NamedTuple):
+    start_time: float  # Negative start times are relative to the end of the match
+    # Negative values 0 - (-0.08) don't get included in the produced recordings
+    fade_time: Optional[float] = None
+    lighting: List[ArenaLighting] = [ArenaLighting('SUN')]
+    luminosity: float = 0.35
+    name: str = ""
+
+    def __repr__(self) -> str:
+        lights_info = [
+            f"({light.light_def}, int={light.intensity}, col={light.colour})"
+            for light in self.lighting
+        ]
+        return (
+            f"<LightingEffect: {self.name!r}, "
+            f"start={self.start_time}, fade={self.fade_time}, "
+            f"lum={self.luminosity}, "
+            f"{', '.join(lights_info)}"
+            ">"
+        )
+
+
+@dataclass
+class LightFade:
+    light: Node
+    remaining_steps: int
+    intensity_step: float
+    colour_step: Tuple[float, float, float]
+    current_intensity: float
+    current_colour: Tuple[float, float, float]
+    effect: ArenaLighting
+
+
+@dataclass
+class LuminosityFade:
+    remaining_steps: int
+    luminosity_step: float
+    current_luminosity: float
+    final_luminosity: float
+
+
+CUE_STACK = [
+    LightingEffect(
+        0,
+        lighting=[ArenaLighting('SUN', intensity=1)],
+        luminosity=0.05,
+        name="Pre-set",
+    ),
+    LightingEffect(0, fade_time=1.5, name="Fade-up"),
+    LightingEffect(
+        -0.1,
```

Comment:
This feels like an unnecessary feature. Perhaps just explicitly saying it's 119.9?

Reply:
This is so that when a custom duration is passed to `run_comp_match` the end of match lighting still works.